projects/chaos/src/relay.c

```diff
@@ -0,0 +1,143 @@
+#include "relay.h"
+
+#include <stdbool.h>
+
+#include "can.h"
+#include "can_ack.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "chaos_events.h"
+#include "event_queue.h"
+#include "fsm.h"
+#include "log.h"
+#include "status.h"
+
+static bool prv_guard(const FSM *fsm, const Event *e, void *context) {
+  CANAckRequest *ack = context;
+  RelayId *id = ack->context;
+  return e->data == *id;
+}
+
+static StatusCode prv_ack_open_callback(CANMessageID msg_id, uint16_t device, CANAckStatus status,
+                                        uint16_t num_remaining, void *context) {
+  RelayId *id = context;
+  if (status != CAN_ACK_STATUS_OK) {
+    event_raise(CHAOS_EVENT_RETRY_RELAY, *id);
+    return STATUS_CODE_OK;
+  }
+
+  event_raise(CHAOS_EVENT_RELAY_OPENED, *id);
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_ack_close_callback(CANMessageID msg_id, uint16_t device, CANAckStatus status,
+                                         uint16_t num_remaining, void *context) {
+  RelayId *id = context;
+  if (status != CAN_ACK_STATUS_OK) {
+    event_raise(CHAOS_EVENT_RETRY_RELAY, *id);
+    return STATUS_CODE_OK;
+  }
+
+  event_raise(CHAOS_EVENT_RELAY_CLOSED, *id);
+  return STATUS_CODE_OK;
+}
+
+#define RELAY_FSM_CFG(name, id)                                                     \
```

Comment:
Can you rename this to something that reflects the setup that you're doing?

Reply:
Done.